Adding a soft limit to the field name length. Closes #33651

alex101101 opened the pull request:
Adds an optional limit to the length of field names, throws an IllegalArgumentException if the limit is breached. Closes #33651

alex101101 committed:
Adding a soft limit to the field name length. Closes #33651

elasticmachine commented:
Pinging @elastic/es-search

cbuescher commented on docs/reference/mapping.asciidoc:
```diff
@@ -36,6 +36,9 @@ treated. Examples of meta-fields include the document's
 A mapping type contains a list of fields or `properties` pertinent to the
 document.
 
+`index.mapping.field.name.length.limit`::
```
Maybe this can be moved to the "Settings to prevent mappings explosion" section since there are already some mapping related settings there, although this is not really a setting that limits the number of fields (and thus is not really that relevant to prevent mapping explosion). I just think while this setting also fits here, its probably not really important enough to appear in this very general section.

If you agree its a good idea to move it I'd put it at the bottom of the refered section and add a short sentence that this setting isn't really something that adresses mappings explosion but might still be useful if you want to limit the field length. I would also state that it usually shouldn't be necessary to set it and the defaults are okay unless users start to add really huge numbers of fields with really long names (in the thousands probably))

cbuescher commented on server/src/main/java/org/elasticsearch/index/mapper/MapperService.java:
```diff
@@ -623,6 +627,23 @@ private void checkDepthLimit(String objectPath, long maxDepth) {
+            .forEach(mapper -> {
+                String name = mapper.simpleName();
+                if (name.length() > maxFieldNameLength) {
+                    throw new IllegalArgumentException("Field name [" + name + "] is too long; limit is length "
```
Maybe rephrase: "the limit is set to ... characters but was..."

cbuescher replied:
Can you also add the index name to the exception message? You can get it through `index().getName()`.

cbuescher commented on server/src/main/java/org/elasticsearch/index/mapper/MapperService.java:
```diff
@@ -101,7 +102,9 @@
-            Setting.longSetting("index.mapping.depth.limit", 20L, 1, Property.Dynamic, Property.IndexScope);
+        Setting.longSetting("index.mapping.depth.limit", 20L, 1, Property.Dynamic, Property.IndexScope);
+    public static final Setting<Long> INDEX_MAPPING_FIELD_NAME_LENGTH_LIMIT_SETTING =
+        Setting.longSetting("index.mapping.field.name.length.limit", Long.MAX_VALUE, 15L, Property.Dynamic, Property.IndexScope);
```
While I think a minimum length of 15 makes much sense, I would probably set the allowed minimum to 1. Some of the other minimum values here also are not very practical, but by setting it to something larger this leads to question like "why not 10, 5, 3?".

alex101101 replied:
Hi Christoph, Thanks for reviewing.  I forgot to mention this when opening the PR but the current check is quite blunt. If a user sets the limit to a low value,
such as 1, a number of templates (from watcher and monitoring plugins) that are set up automatically on server startup will fail. I can look into finding a method to 
skip the length check for system created mappers if you think that is needed? Also it checks `MetaDataFieldMapper`s field name length, I could put an `instanceof` check in to skip these also if needed.

cbuescher replied:
This sounds strange. I was under the impression this setting (if set at all), only applies to a certain index. Are you saying that system-level indices and templates (like .watcher etc...) would be effected by this even though if a user changes this for index `fooBar` only?

alex101101 replied:
Thanks that makes sense, I was testing it incorrectly.

cbuescher commented on server/src/test/java/org/elasticsearch/index/mapper/MapperServiceTests.java:
```diff
@@ -323,4 +323,112 @@ public void testDefaultMappingIsRejectedOn7() throws IOException {
+                .endObject()
+            .endObject().endObject()));
+
+        DocumentMapper documentMapper = mapperService.merge("type", mapping, MergeReason.MAPPING_RECOVERY);
```
Great to also include this recovery test, I wouldn't have remembered this should be handled differently grom normal mapping updates.

cbuescher reviewed (changes requested):
Hi @alex101101, thanks for working on this issue and opening the PR, I think it looks great and left only a few minor suggestions to maybe change a bit of the documentation and improve the error message a small bit. Let me know if you disagree with any of my comments, otherwise I think with a small update we can get this in quickly.

alex101101 committed:
Peer Review Comment pt1

alex101101 commented:
Hi @cbuescher, Thanks for reviewing this.  I have addressed your comments except for one of them which I have left a reply to and was hoping for some clarification on/a second opinion.

jasontedor commented on docs/reference/mapping.asciidoc:
```diff
@@ -96,7 +96,14 @@ causing a mapping explosion:
     100 objects within a nested field, will actually create 101 documents, as
     each nested object will be indexed as a separate hidden document.
-
+`index.mapping.field.name.length.limit`::
```
I think this should be named: `index.mapping.field_name_length.limit`. We don't need a namespace for every component.

jasontedor reviewed (changes requested):
I left one comment about the setting name.

cbuescher commented on docs/reference/mapping.asciidoc:
```diff
@@ -96,7 +96,14 @@ causing a mapping explosion:
+`index.mapping.field.name.length.limit`::
+    Setting for the maximum length of a field name. The default value is
+    Long.MAX_VALUE (no limit).
+    *Note*
```
Thanks for moving this, small follow up nit-pick: I don't really like adding the two points mentioned here as special "Note" with sub-bulltest here beacuse I think it draws too much attention to an otherwise quite marginal setting. I think it should be enough to either just inline the two additional sentences into the previous paragraph and/or put the first sentence in front of the setting description.

alex101101 committed:
Peer Review Comment pt2

alex101101 commented:
Hi @cbuesche,

Cheers. I have updated the PR according to your (and @jasontedor's) latest comments.

cbuescher commented:
@elasticmachine test this please

cbuescher reviewed (approved):
Thanks a lot for the updates @alex101101, this looks good to me. Will run the tests and merge if everything passes, unless @jasontedor wants to take another look.

jasontedor commented:
@cbuescher If you're good with it, I'm good with it. My only comment was a drive-by comment about the naming. :innocent:

Commit:
Merge branch 'master' into soft-limit-field-name

cbuescher commented:
@alex101101 fyi I merged in master due to some backwards comatibility test errors, hope those should be resolved by this
@elasticmachine test this please

cbuescher commented:
@alex101101 the tests uncovered a small issue the new setting causes with the new CCR feature in x-pack. It seems settings need to be classified as "replicated or not replicated". You can run the failing test in isolation with

```
./gradlew :x-pack:plugin:ccr:unitTest -Dtests.seed=E402D2B2896AA1B3 -Dtests.class=org.elasticsearch.xpack.ccr.action.TransportResumeFollowActionTests -Dtests.method="testDynamicIndexSettingsAreClassified" -Dtests.security.manager=true -Dtests.locale=es-BO -Dtests.timezone=America/Anchorage -Dcompiler.java=12 -Druntime.java=8
```
I think you can just add the new setting as a setting that needs to be replicated in TransportResumeFollowActionTests like this. At least thats how other similar index mapping settings are treated:
```
         replicatedSettings.add(MapperService.INDEX_MAPPING_NESTED_FIELDS_LIMIT_SETTING);
         replicatedSettings.add(MapperService.INDEX_MAPPING_DEPTH_LIMIT_SETTING);
+        replicatedSettings.add(MapperService.INDEX_MAPPING_FIELD_NAME_LENGTH_LIMIT_SETTING);
```

If you like I can also push that fix. Let me know how you decide so I can re-run the test suite again.

alex101101 commented:
Hi @cbuescher,

Thanks would be great if you can push the fix if you already have it. From what I can see, I agree, this setting probably needs to be replicated as if it is different between clusters it could affect index replication.

Commit:
Add new setting to replicated settings in TransportResumeFollowActionTests

cbuescher commented:
@elasticmachine test this please

cbuescher commented:
@alex101101 test passed, I merged to master and will backport this PR to the 7.1 branch. Thanks a lot for this contribution, much appreciated.

alex101101 commented:
Great. Thanks very much for your help getting this PR merged @cbuescher
